create data path for s390 FDR backups if !exists

mutable-dan opened the pull request:
if the path for $FDRUPSTREAM_DATA_PATH does not exist, then create it.
This change is FDR specific for s390 and is only in effect if ZVM_NAMING is "Y"

#### Relax-and-Recover (ReaR) Pull Request Template

Please fill in the following items before submitting a new pull request:

##### Pull Request Details:

* Type: Bug Fix

* Impact:  Normal

Description:
If the data destination dir does not exist, then create it

jsmeix commented on usr/share/rear/layout/save/FDRUPSTREAM/Linux-s390/990_copy_disklayout_file.sh:
```diff
@@ -24,6 +24,7 @@ opath=$( output_path $scheme $path )
 if [[ "$ZVM_NAMING" == "Y" && "$ARCH" == "Linux-s390" ]] ; then 
       VM_UID=$(vmcp q userid |awk '{ print $1 }')
       LogPrint "s390 dislayout.conf will be saved as $(opath)/$VM_UID.disklayout.conf"
+      mkdir -p ${opath}/"$VM_UID"
```
The command
```
LogPrint "s390 dislayout.conf will be saved as $(opath)/$VM_UID.disklayout.conf"
```
cannot work because `$(opath)` is command substitution as follows:
```
# opath=/tmp//path/to/somewhere

# VM_UID=vm-UID

# echo "s390 dislayout.conf will be saved as $(opath)/$VM_UID.disklayout.conf"
opath: command not found
s390 dislayout.conf will be saved as /vm-UID.disklayout.conf
```
Usually you do not need the complicated `${var1}/${var2}` syntax
unless the character in between is one of the allowed characters
in a bash variable name, cf. "man bash"
```
word
  A sequence of characters considered
  as a single unit by the shell.
  Also known as a token.

name
  A word consisting only of alphanumeric characters
  and underscores, and beginning with an alphabetic
  character or an underscore.
  Also referred to as an identifier.
```
So `$opath/$VM_UID` is simpler and in the end more fail safe.
Accordingly I would suggest to do it this way:
```
VM_UID=$( vmcp q userid |awk '{ print $1 }' )
LogPrint "s390 dislayout.conf will be saved as $opath/$VM_UID.disklayout.conf"
mkdir -p $opath/$VM_UID
```

That seems to work at least for me:
```
# opath=/tmp/path/to/somewhere

# VM_UID=vm-UID

# echo "s390 dislayout.conf will be saved as $opath/$VM_UID.disklayout.conf"
s390 dislayout.conf will be saved as /tmp/path/to/somewhere/vm-UID.disklayout.conf

# mkdir -vp $opath/$VM_UID
mkdir: created directory '/tmp/path'
mkdir: created directory '/tmp/path/to'
mkdir: created directory '/tmp/path/to/somewhere'
mkdir: created directory '/tmp/path/to/somewhere/vm-UID'
```

jsmeix commented:
@mutable-dan 
thank you for your continuous testing of ReaR on IBM Z
and for your fixes and adaptions!

mutable-dan commented:
You are welcome:
LogPrint... oops.   fixed
simplified the constructs as you recommended in this file and two other paces

usr/share/rear/layout/save/FDRUPSTREAM/Linux-s390/990_copy_disklayout_file.sh
usr/share/rear/output/default/950_copy_result_files.sh
usr/share/rear/pack/GNU/Linux/900_create_initramfs.sh

jsmeix commented on usr/share/rear/layout/save/FDRUPSTREAM/Linux-s390/990_copy_disklayout_file.sh:
```diff
@@ -23,7 +23,17 @@ opath=$( output_path $scheme $path )
+            VM_UID="DEFAULT"
+      fi
+      if [[ -z $opath ]] ; then
+            LogPrint "Warning: output path is not set, please check OUTPUT_URL in local.conf."
```
Scripts that are run during "rear mkrescue/mkbackup"
(e.g. scripts in the `layout/save` stage - cf. the `rear -s mkrescue` output)
are normally run non-verbose so `LogPrint` messages are normally
not shown to the user while "rear mkrescue/mkbackup" runs.

Therefore `LogPrint "Warning..."` messages are normally
not shown to the user while "rear mkrescue/mkbackup" runs.

To show such messages in any case to the user use `LogPrintError "..."`
cf. the `LogPrintError` function comment in lib/_input-output-functions.sh
https://github.com/rear/rear/blob/master/usr/share/rear/lib/_input-output-functions.sh#L445

In general regarding `Warning` messages see
https://blog.schlomo.schapiro.org/2015/04/warning-is-waste-of-my-time.html

When the subsequent code does not work well with an empty `$opath`
e.g. when
```
cp $v $DISKLAYOUT_FILE /$VM_UID.disklayout.conf
```
is not a supported behaviour, I think it is better for the user
to make a clear decision here and error out e.g. with something like
```
Error "Output path is empty (OUTPUT_URL='$OUTPUT_URL')"
```
The `Error` function shows in which script it was called like
```
# usr/sbin/rear mkrescue
ERROR: Output path is empty (OUTPUT_URL='nfs://192.168.122.1/nfs')
Some latest log messages since the last called script 990_copy_disklayout_file.sh
  ...
```
so the user knows where to look if he was hit by a false error exit
and he needs to adapt the code that causes the false error exit.

jsmeix commented on usr/share/rear/output/default/950_copy_result_files.sh:
```diff
@@ -72,16 +72,28 @@ case "$scheme" in
+
             if [[ "$ZVM_NAMING" == "Y" && "$ARCH" == "Linux-s390" ]] ; then 
+               if [[ -z $opath ]] ; then
+                  LogPrint "Warning: output path is not set, please check OUTPUT_URL in local.conf."
```
See my comment above
https://github.com/rear/rear/pull/2320/files#r371245630

jsmeix commented:
@rear/contributors 
to be more on the safe side I would appreciate
a second review from another ReaR contributor

mutable-dan commented:
good point on the error vs warning.  made the change

mutable-dan commented:
with the approvals done, I will merge the request.
thank you

jsmeix commented:
@mutable-dan 
thank you for your continuous testing and your improvements
in particular for external/third-party backup software!

Usually we do not have such software so we depend
on contributions from users of such software.

pcahyna commented on usr/share/rear/layout/save/FDRUPSTREAM/Linux-s390/990_copy_disklayout_file.sh:
```diff
@@ -23,8 +23,17 @@ opath=$( output_path $scheme $path )
+
+      LogPrint "s390 disklayout.conf will be saved as $opath/$VM_UID.disklayout.conf"
+      mkdir -pv $opath
+      cp $v $DISKLAYOUT_FILE $opath/$VM_UID.disklayout.conf || Error "Failed to copy disklayout.conf ($DISKLAYOUT_FILE) to opath/$VM_UID.disklayout.conf"
```
`opath` should be `${opath} in the Error message.

jsmeix replied:
Fixed via
https://github.com/rear/rear/commit/7bcad6b517aad2fcaf42d2a12464b42b5508a856
plus much more code cleanup that "just happened by the way" as I noticed it ;-)
